CONTRIBUTING.md

```diff
@@ -0,0 +1,437 @@
+- Builders can be used instead of constructors or factory methods with large amounts of parameters.
+
+### API class modifiers and member visibility
+
+- Classes in Fabric API should be `final` classes unless the class exposed in the API is explicitly meant for extension.
+- `private` constructors should be used in API classes unless the class is explicitly meant to be instantiated.
+    - This only applies to modder-facing classes, i.e. classes in the `net.fabricmc.fabric.api` subpackage. See below for package structure.
+    - They should be placed at the very bottom of the class to not hurt readability.
+- Access modifiers for fields and methods should be as strict as possible.
+    - If a method is intended to only be for use by mods implementing an api, a `protected` method should suffice.
+
+### Annotations
+- Nullable members, parameters or return values must be annotated with the `@Nullable` annotation.
+    - Any member, parameter or return value that is not marked as nullable can be assumed to be nonnull. The `@NotNull` annotation should never be used, it is implicit.
+    - Introducing custom types might be appropriate (no need to specify `List<@Nullable String>` everywhere).
+- Deprecated API members should have the `@Deprecated` annotation.
+    - Avoid specifying `forRemoval = true`, the functionality will be supported as long as possible.
+        - Exceptions are made for experimental APIs, where deprecated for removal functionality might be removed after sufficient time has passed.
+    - A javadoc `@deprecated` comment should be added to deprecated members with migration instructions.
+- Experimental API classes should all have the `@ApiStatus.Experimental` annotation.
+    - The annotation is not required on individual members, unless the class itself is not experimental.
+- `@ApiStatus.NonExtendable` should be used for API interfaces or classes that modders must not implement or override, but can't be `final` for some reason.
+    - Adding methods to such interfaces or classes is not a breaking change.
+- `@ApiStatus.Internal` is recommended for implementation classes to avoid IDE autocompletion suggesting them. There should be no need for internal methods or classes in the public API package.
+
+### Naming
+- All names should follow the [Yarn naming standards](https://github.com/FabricMC/yarn/blob/HEAD/CONVENTIONS.md).
+- If a class only contains getter methods, the `get` prefix may be omitted for methods. The `get` prefix may also be omitted where it is not appropriate.
+- Accessor mixins should be named **Target**Accessor, other mixins should be named **Target**Mixin, where `Target` is the target class name. More details in the mixin section below.
+
+### Events
+
+- Events should not be used if there is only one subscriber, like a handler in a registered unique namespace.
+- Events should be produced and fully usable without object allocation.
```

Comment:
```suggestion
- Events should be produced and fully usable without creating data holder objects for inputs.
```
Events sometimes do need to allocate objects, for example when they use `TypedActionResult`

Reply:
It's a general design guideline. While it mainly applies for inputs, it could also apply for outputs in some cases (e.g. pass a `Consumer` parameter instead of returning a `List`).

Reply:
The current version to me reads like "you should never allocate objects when designing events", which does not sound like a soft general guideline